File: Depth-First Search/Combinations/Numbers/LET_no78_Subsets.py
# The same as LeetCode no78. Subsets

class Solution:
    """
    @param nums: A set of Numbers
    @return: A list of lists
    """

    def subsets(self, nums):
        nums.sort()
        result = []
        self.dfs(nums, 0, [], result)
        return result

    def dfs(self, nums, start, path, result):
        result.append(path[:])

        for i in range(start, len(nums)):
            path.append(nums[i])
            self.dfs(nums, i + 1, path, result)
            path.pop()

File: Depth-First Search/Combinations/Numbers/LET_no78_Subsets.cpp
class Solution {
private:
    int n;
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> temp;
        helper(nums, 0, temp, res);
        return res;
    }

    void helper(vector<int>& nums, int start, vector<int>& temp, vector<vector<int>>& res) {
        res.push_back(temp);

        for (int i = start; i < n; i++) {
            temp.push_back(nums[i]);
            helper(nums, i + 1, temp, res);
            temp.pop_back();
        }
    }
};